pkg/controller/cloud/pvlcontroller.go

```diff
@@ -0,0 +1,373 @@
+	for i := 0; i < threadiness; i++ {
+		// runWorker will loop until "something bad" happens.  The .Until will then rekick the worker
+		// after one second
+		go wait.Until(pvlc.runWorker, time.Second, stopCh)
+	}
+
+	// wait until we're told to stop
+	<-stopCh
+}
+
+func (pvlc *PersistentVolumeLabelController) runWorker() {
+	// hot loop until we're told to stop.  processNextWorkItem will automatically wait until there's work
+	// available, so we don't worry about secondary waits
+	for pvlc.processNextWorkItem() {
+	}
+}
+
+// processNextWorkItem deals with one key off the queue.  It returns false when it's time to quit.
+func (pvlc *PersistentVolumeLabelController) processNextWorkItem() bool {
+	// pull the next work item from queue.  It should be a key we use to lookup something in a cache
+	keyObj, quit := pvlc.queue.Get()
+	if quit {
+		return false
+	}
+	// you always have to indicate to the queue that you've completed a piece of work
+	defer pvlc.queue.Done(keyObj)
+
+	key := keyObj.(string)
+	// do your work on the key.  This method will contains your "do stuff" logic
+	err := pvlc.syncHandler(key)
+	if err == nil {
+		// if you had no error, tell the queue to stop tracking history for your key.  This will
+		// reset things like failure counts for per-item rate limiting
+		pvlc.queue.Forget(key)
+		return true
+	}
+
+	// there was a failure so be sure to report it.  This method allows for pluggable error handling
+	// which can be used for things like cluster-monitoring
+	utilruntime.HandleError(fmt.Errorf("%v failed with : %v", key, err))
+
+	// since we failed, we should requeue the item to work on later.  This method will add a backoff
+	// to avoid hotlooping on particular items (they're probably still not going to work right away)
+	// and overall controller protection (everything I've done is broken, this controller needs to
+	// calm down or it can starve other useful work) cases.
+	pvlc.queue.AddRateLimited(key)
+
+	return true
+}
+
+// AddLabels adds appropriate labels to persistent volumes and sets the
+// volume as available if successful.
+func (pvlc *PersistentVolumeLabelController) addLabels(key string) error {
+	_, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		return fmt.Errorf("error getting name of volume %q to get volume from informer: %v", key, err)
+	}
+	volume, err := pvlc.volumeLister.Get(name)
+	if errors.IsNotFound(err) {
+		return nil
+	} else if err != nil {
+		return fmt.Errorf("error getting volume %s from informer: %v", name, err)
+	}
+
+	return pvlc.addLabelsToVolume(volume)
+}
+
+func (pvlc *PersistentVolumeLabelController) addLabelsToVolume(vol *v1.PersistentVolume) error {
+	var volumeLabels map[string]string
+
+	// Only add labels if in the list of initializers
+	if needsInitialization(vol.Initializers, initializerName) {
+		if vol.Spec.AWSElasticBlockStore != nil {
+			labels, err := pvlc.findAWSEBSLabels(vol)
```

Comment:
@rrati How long is this call going to take? We might enforce a global timeout on all initializers.

Reply:
@rrati I put the `do-not-merge` label on the PR for this question because it might affect the evolving of initializers.

Reply:
This is the same logic that is in the admission plugin.  It will be going away in a future revision of this controller.  The call is very fast in my testing, but I haven't hit any instances where the API won't respond.

Reply:
@caesarxuchao what happens if this initializer times out? Is the object considered uninitializable?

Reply:
We perhaps will let the apiserver or garbage collector delete the object.

> The call is very fast in my testing

How fast is the API call? Less than 1s? Could you do caching?

Reply:
The API call itself is <<1s in my testing.  However, that number could go up if the cloud provider is having issues, and I don't have any numbers for that.  I never ran into a case where that occurred.

Reply:
<<1s sounds good. Thanks.